rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md

```diff
@@ -0,0 +1,495 @@
+---
+authors: Marco Dinis (marco.dinis@goteleport.com)
+state: draft
+---
+
+# RFD 999 - OIDC IdP for AWS RDS DBs discovery
+
+## Required Approvals
+
+* Engineering: @ryanclark && @r0mant
+* Product: @xinding33 || @klizhentas
+* Security: @reedloden
+
+## What
+
+Discover AWS RDS Databases without manual configuration.
+
+#### Goals
+
+Easily discover AWS RDS Databases of an account.
+
+#### Non-goals
+
+Having a real OIDC IdP is out of scope.
+That initiative can be tracked [here](https://github.com/gravitational/teleport/issues/20967).
+
+Automatically onboard AWS RDS Databases into Teleport.
+
+Discovering other AWS resources is out of scope.
+
+## Terminology
+
+* OIDC: OpenID Connect is a protocol that works on top of OAuth2 and provides identity information.
+* IdP: Identity Provider is a system that manages identities.
+* JWT: JSON Web Token is a json encoded token that contains a set of claims signed by an entity.
+* JWKS: JSON Web Key Set is a set of public keys that IdP consumers must use to validate a signed JWT.
+
+## Why
+
+Our current discovery process for AWS resources requires multiple steps before the user gets to the "aha moment".
+Some of those steps are a little complicated and require the user to know some Teleport specific internals.
+
+We can reduce the number of steps and remove most of the Teleport specific configuration by creating an OIDC IdP in Teleport.
+
+## How
+
+AWS allows the [set up](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) of an IdP using OIDC providers.
+Each provider has a set of roles, which limit their access.
+
+Turning Teleport into an OIDC IdP, and asking the user to create a new OIDC IdP that consumes the Teleport OIDC endpoints, allows us to easily call AWS API endpoints to, for instance, list RDS instances and their details.
+
+When configuring the provider, we'll need an AWS Role which Teleport uses to issue API Calls.
+
+To store the configuration above, we'll create a new resource Kind: `Integration`.
+It'll leverage the `subkind` prop to distinguish future integrations.
+
+```yaml
+kind: integration
+subkind: aws-oidc
```

Comment:
Nit: Do we need `-oidc` suffix?

Reply:
I'm trying to leave space/pattern to have other types of AWS integrations (using secrets or using an ec2 instance).

Should I remove the `-oidc` anyway?